### PR TITLE
Add |wpt| command frontend

### PR DESCRIPTION
**Pull request by jgraham**
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/6237)

### CODE REVIEW

**Comment by jgraham** [comment]
Not quite done yet (missing docs), but should work in the current state.

**Comment by bobholt** [comment]
One apparent bug, and yes, docs are welcome.

Realizing this may be a bit of a naive question: Why this approach and not something like `setup.py` where `wpt` aliases run/install/etc. commands in `setup.py`?

---

Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[tools/lint/lint.py, line 768 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6237#-KmaQeeUAlHGGy6So8Y7:-KmaQeeUAlHGGy6So8Y8:bhor5c9) ([raw file](https://github.com/w3c/web-platform-tests/blob/98d95d37910584750bcdfb1f69e1d76a4c668d2e/tools/lint/lint.py#L768)):*
> ```Python
> 
> if __name__ == "__main__":
>     args = create_parser.parse_args()
> ```

Looks like this should be `create_parser().parse_args()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6237#-:-KmaS48jQi81P4wnz1oL:bilu3z2)*

**Comment by jgraham** [comment]
AFAIK if you create a setup.py then the user has to create their own virtualenv and run python setup.py install &c. Making it possible to run everything from a single command without any setup apart from cloning the repo is very desirable (we could consider adding a command to install the virtualenv tool as well, for example). This approach is based heavily — in spirit if not im implementation — on Mozilla's `mach` command, which has the same advantage of not requiring any python-specific knowledge to e.g. start a build.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242786715)

# Firefox (nightly) #
Testing web-platform-tests at revision 9f3b0460e628fa586413abe7b693f09b66a96390
('python', 'manifest') exited with return code 1
<function run at 0x7fb8ac309668>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242786716)

# Sauce (safari) #
Testing web-platform-tests at revision 9f3b0460e628fa586413abe7b693f09b66a96390
('python', 'manifest') exited with return code 1
<function run at 0x7f4b244ea668>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242786726)

# Chrome (unstable) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242786727)

# Sauce (MicrosoftEdge) #

**Comment by jugglinmike** [comment]
Happy to see further consolidation happening. This patch omits at least three files: `/wpt`, `/tools/lint/commands.json`, `/tools/manifest/commands.json`. I recommend naming those "commands" files `wpt_commands.json` to give a hint about their purpose in the otherwise-unrelated directories. I'll hold off on further review until I'm able to play with this locally.

File: tools/wpt/run.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# wpt command:run:main
```
Comment (jugglinmike) [inline tools/wpt/run.py]: I'm having trouble understanding the purpose of this comment. Is it a vestige of some previous implementation strategy? Or am I missing something about the current patch?

File: tools/manifest/update.py [hunk 2]
```diff
@@ -81,24 +82,16 @@ def find_top_repo():
...
-        opts.tests_root = tests_root
+def main():
+    opts = create_parser().parse_args()
 
-    if opts.path is None:
-        opts.path = os.path.join(opts.tests_root, "MANIFEST.json")
+    run(**vars(opts))
 
-    update_from_cli(**vars(opts))
+print run
```
Comment (jugglinmike) [inline tools/manifest/update.py]: Errant `print` statement

File: tools/browserutils/commands.json [hunk 3]
```diff
@@ -0,0 +1,2 @@
+{"install-browser": {"path": "install.py", "script": "run", "parser": "get_parser", "help": "Install browser components",
```
Comment (jugglinmike) [inline tools/browserutils/commands.json]: `wpt install-browser firefox browser` looks a bit redundant, while `wpt
install-browser firefox webdriver` is a little misleading. What do you think
about calling this sub-command `install` instead?

File: tools/wpt/run.py [hunk 4]
```diff
@@ -295,10 +310,7 @@ def main():
     venv.install_requirements(os.path.join(wpt_root, "tools", "wptrunner", "requirements.txt"))
     venv.install("requests")
```
Comment (jugglinmike) [inline tools/wpt/run.py]: Is this still necessary?

File: tools/wpt/commands.json [hunk 5]
```diff
@@ -0,0 +1,2 @@
+{"run": {"path": "run.py", "script": "run", "parser": "create_parser", "help": "Run tests in a browser",
+         "virtualenv": true, "install": ["requests"], "requirements": ["../wptrunner/requirements.txt"]}}
```
Comment (jugglinmike) [inline tools/wpt/commands.json]: If the "requests" module is necessary to use the `wptrunner` module, why isn't it listed in `wptrunner/requirements.txt`?

File: tools/wpt/run.py (hunk 4 above)
Comment (jgraham) [inline tools/wpt/run.py]: Not necessary, but it allows the file to be run directly from the command line, which seems somewhat useful to keep for now.

File: tools/wpt/commands.json (hunk 5 above)
Comment (jgraham) [inline tools/wpt/commands.json]: It's necessary to install the browser/webdriver/prefs/etc. so it's a requirement for `wpt run` but not for wptrunner itself.

File: tools/wpt/wpt.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                }
+    return rv
+
+
+def parse_args(commands):
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--venv", action="store", help="Path to an existing virtualenv to use")
+    parser.add_argument("--debug", action="store_true", help="Run the debugger in case of an exception")
+    subparsers = parser.add_subparsers(dest="command")
+    for command, props in commands.iteritems():
```
Comment (jgraham) [inline tools/wpt/wpt.py]: Shrug. I guess I ought to make this work with Python 3, but wptrunner doesn't, for example, so it won't be *super* useful.

**Comment by jgraham** [comment]
This is ready and waiting for review.

**Review by jugglinmike — COMMENTED** [review]
The `README.md` file contains a few lingering references to `wptrun`: `wptrun --help`, `wptrun --wptrunner-help` `wptrun --binary=path product` (I don't think this last one is supported by `wpt`, though).

The `lint.whitelist` file includes a reference to the `wptrun.py` file that this patch renames. The tests pass, so I guess this can be removed (rather than updated).

Would you mind removing the `wptrun, `lint` and `whitelist` executables from the top level of the project?

**Comment by jgraham** [comment]
I want to leave the executables for now in case downstream consumers are making use of them in scripts or integrations. They can be removed in a followup once this has been announced.

**Comment by jgraham** [comment]
@jugglinmike updated this.

**Comment by jugglinmike** [comment]
Got it. This latest version is rebased on `master`, but I can't execute `wptrun` from `master` at the moment, so I'll need to work that out before I can continue the review here. See https://github.com/w3c/web-platform-tests/pull/6375#issuecomment-311804976

**Comment by jugglinmike** [comment]
A couple of the documented options for the `run` sub-command do not work.

Specifying `--binary` produces the following output:

    $ ./wptrun --binary /opt/mozilla/firefox-nightly/firefox firefox url/a-element.html 
    usage: wptrun [-h] [--yes] [--wptrunner-help] product [tests [tests ...]] ...
    wptrun: error: unrecognized arguments: --binary

Specifying `--wptrunner-help` produces the following output:

    $ ./wpt run --wptrunner-help
    Traceback (most recent call last):
      File "./wpt", line 5, in <module>
        wpt.main()
      File "/home/mike/projects/bocoup/google-wpt/web-platform-tests/tools/wpt/wpt.py", line 100, in main
        kwargs = vars(parser.parse_args(command_args))
      File "/usr/lib/python2.7/argparse.py", line 1701, in parse_args
        args, argv = self.parse_known_args(args, namespace)
      File "/usr/lib/python2.7/argparse.py", line 1733, in parse_known_args
        namespace, args = self._parse_known_args(args, namespace)
      File "/usr/lib/python2.7/argparse.py", line 1939, in _parse_known_args
        start_index = consume_optional(start_index)
      File "/usr/lib/python2.7/argparse.py", line 1879, in consume_optional
        take_action(action, args, option_string)
      File "/usr/lib/python2.7/argparse.py", line 1807, in take_action
        action(self, namespace, argument_values, option_string)
      File "/home/mike/projects/bocoup/google-wpt/web-platform-tests/tools/wpt/run.py", line 37, in __call__
        from wptrunner import wptcommandline
      File "/home/mike/projects/bocoup/google-wpt/web-platform-tests/tools/wptrunner/wptrunner/wptcommandline.py", line 9, in <module>
        import wpttest
      File "/home/mike/projects/bocoup/google-wpt/web-platform-tests/tools/wptrunner/wptrunner/wpttest.py", line 4, in <module>
        import mozinfo
    ImportError: No module named mozinfo

...although I receive similar errors when using `wptrun` as defined in `master` today.

**Comment by gsnedders** [comment]
There's a lot of new untested code here, especially given @jugglinmike found issues earlier. :/

**Comment by jgraham** [comment]
Closing this in favour of #6421 that ended up fiing some issues here.

**Pull request closed without merging**